### PR TITLE
Recompose security-openid-connect-web-authentication.adoc to Diataxis framework

### PR DESCRIPTION
**Pull request by sheilamjones**
Recompose the guide USING OPENID CONNECT (OIDC) TO PROTECT WEB APPLICATIONS USING AUTHORIZATION CODE FLOW (security-openid-connect-web-authentication.adoc) into the Diataxis framework. 

The focus of this PR is to move and restructure the content into the topics. Further iterations are needed to further recomposition and style/editing changes

The following updates were made:
- Moved the overview, OIDC diagram, and all sections under ==Reference guide into a new file called xref:security-oidc-code-flow-authentication-concept.adoc 
- Moved all elements in the ==Quickstart section into a new file called xref:security-oidc-code-flow-authentication-tutorial.adoc
- Updated and fixed links to point to new topics and moved content- 
- Completed minor style changes and identified further enhancements needed
- Fixed existing links to the content that was moved.

Closes: https://issues.redhat.com/browse/QDOCS-85 
(and subtasks  https://issues.redhat.com/browse/QDOCS-109 and  https://issues.redhat.com/browse/QDOCS-110)

### CODE REVIEW

**Comment by sberyozkin** [comment]
Good evening @sheilamjones Thanks for this PR, I'd consider naming it a `security-protect-web-authentication-tutorial`, simpler name, only code flow is supported in any case but we can discuss it next week, FYI, I've seen https://github.com/quarkusio/quarkus/pull/30456/files, but haven't started reviewing it as I guess the will be some links updated to link to these new files, etc.

Enjoy the weekend

**Comment by sberyozkin** [comment]
Hi Sheila, should this PR be combined with another PR you opened ? I'd like to see how both new files are linked to, you probably need to rebase once @michelle-purcell 's PR is merged, as the OIDC intro section will be in the new auth mechanisms doc.
Please also simplify/shorten the name of the new file as proposed in the prev comment,  the shorter the final URL segment the better IMHO
Thanks

**Comment by sheilamjones** [comment]
@sberyozkin: Thanks for your comments above. I have renamed the new tutorial file to shorten it per your suggestions. In this PR, the Quickstart content from the original file (security-openid-connect-web-authentication.adoc) file has been removed from there and recomposed into this new tutorial file. I also rebased to pull in Michelle's latest changes.  
Note, further styling needs to be done in the next iteration of updates. 
There is a separate PR (https://github.com/quarkusio/quarkus/pull/30456/files) to handle the recomposing of the conceptual and reference sections of (security-openid-connect-web-authentication.adoc)

File: docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+   }
+}
+----
+
+This endpoint has ID, access, and refresh tokens injected.
+It returns a `preferred_username` claim from the ID token, a `scope` claim from the access token, and also a refresh token availability status.
+
+Note that you do not have to inject the tokens - it is only required if the endpoint needs to use the ID token to interact with the currently authenticated user or use the access token to access a downstream service on behalf of this user.
+
+// SJ: TO DO - update link to point to new reference guide. For more information, see <<access_id_and_access_tokens,Access ID and Access Tokens>> section.
```
Comment (rolfedh) [inline docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc:143]: You're meaning to commit this comment, right?

File: docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+If everything is working as expected, you are redirected to the Keycloak server to authenticate.
+
+To authenticate to the application, type the following credentials when at the Keycloak login page:
+
+* Username: *alice*
+* Password: *alice*
+
+After clicking the `Login` button, you are redirected back to the application.
+
+// Sheila, JAN 23: TO DO - Update link here to point to new concept topic:For more information about writing the integration tests that depend on `Dev Services for Keycloak`, see the <<integration-testing-keycloak-devservices, Dev Services for Keycloak>> section.
```
Comment (rolfedh) [inline docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc:243]: You're meaning to commit this comment, right?

**Review by rolfedh — APPROVED** [review]
LGTM! Nice work.

File: docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc [hunk 3]
```diff
@@ -0,0 +1,264 @@
+[id="security-protect-web-authentication-tutorial"]
```
Comment (michelle-purcell) [inline docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc:1]: Should the ID match the new file name as per the following?

File: docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+= Protect a web application by using OIDC authorization code flow mechanism
+include::_attributes.adoc[]
+:categories: security
```
Comment (michelle-purcell) [inline docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc:4]: Consider adding `web` also. Some of the other web-related security topics also use this category too.
```suggestion
:categories: security,web
```

File: docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+// Sheila, JAN 23: TO DO - Update link here to point to new concept topic:For more information about writing the integration tests that depend on `Dev Services for Keycloak`, see the <<integration-testing-keycloak-devservices, Dev Services for Keycloak>> section.
+
+== Summary
+
+Congratulations!
+You have learned how to set up and use the OIDC authorization code flow mechanism to protect and test application HTTP endpoints.
+After you have completed this tutorial, explore some of the other security mechanisms in Quarkus.
+
+* xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer authentication]
+* xref:security-overview-concept.adoc[Quarkus Security]
```
Comment (michelle-purcell) [inline docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc:252]: The page title was recently changed to `Quarkus Security overview`. 
```suggestion
* xref:security-overview-concept.adoc[Quarkus Security overview]
```

File: docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc [hunk 6]
```diff
@@ -0,0 +1,264 @@
...
+
+== References
+//TO DO, Sheila: update this link when other concept PR is merged - update* xref:security-oidc-web-authentication-concept.adoc[OIDC authorization code flow mechanism]
+* https://www.keycloak.org/documentation.html[Keycloak Documentation]
+* https://openid.net/connect/[OpenID Connect]
+* https://tools.ietf.org/html/rfc7519[JSON Web Token]
+* xref:security-openid-connect-client-reference.adoc[OpenID Connect and OAuth2 Client and Filters Reference Guide]
+* xref:security-openid-connect-dev-services.adoc[Dev Services for Keycloak]
+* xref:security-jwt-build.adoc[Sign and encrypt JWT tokens with SmallRye JWT Build]
+* xref:security-overview-concept.adoc#oidc-jwt-oauth2-comparison[Summary of Quarkus OIDC, JWT and OAuth2 features]
```
Comment (michelle-purcell) [inline docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc:263]: This target page also got recomposed earlier today too and got caught in the security doc PR crossfire!:fire_extinguisher: 

```suggestion
* xref:security-authentication-mechanisms-concept.adoc#oidc-jwt-oauth2-comparison[Choosing between OpenID Connect, SmallRye JWT, and OAuth2 authentication mechanisms]
```

File: docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc [hunk 7]
```diff
@@ -37,236 +37,6 @@ For information about multitenant support, see xref:security-openid-connect-mult
...
-In order to authenticate to the application you should type the following credentials when at the Keycloak login page:
-
-* Username: *alice*
-* Password: *alice*
-
-After clicking the `Login` button you should be redirected back to the application.
-
-Please also see the <<integration-testing-keycloak-devservices, Dev Services for Keycloak>> section below about writing the integration tests which depend on `Dev Services for Keycloak`.
-
 == Reference Guide
```
Comment (michelle-purcell) [inline docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc:40]: Probably a comment that is outside the scope of this PR restructure:

```suggestion
== References
```
Made me wonder if the doc _templates are missing an optional **References** section? Perhaps one for the Contributing guide...

**Review by michelle-purcell — APPROVED** [review]
@sheilamjones - Nice job :+1: Just a few very minor comments. 
Some might be slightly outside the scope of this iteration.

**Comment by sberyozkin** [comment]
Hi @sheilamjones 

Thanks for the updates, 

>https://github.com/quarkusio/quarkus/pull/30456/files

My question is, should #30456 and this PR become a single PR ? Both PR affect the same document, and it is not easy to see how this single `security-openid-connect-web-authentication` is impacted

**Comment by sheilamjones** [comment]
Thanks @sberyozkin. I shall close PR https://github.com/quarkusio/quarkus/pull/30456 and move those changes under this single PR.

File: docs/src/main/asciidoc/security-authentication-mechanisms-concept.adoc [hunk 8]
```diff
@@ -154,7 +154,7 @@ For more information about OIDC authentication and authorization methods you can
 |====
 |OIDC topic |Quarkus information resource
 |Bearer Token authentication mechanism|xref:security-oidc-bearer-authentication-concept.adoc[OIDC Bearer authentication]
-|Authorization Code Flow authentication mechanism|xref:security-openid-connect-web-authentication.adoc[OpenID Connect (OIDC) authorization code flow mechanism]
+|Authorization Code Flow authentication mechanism|xref:security-oidc-web-authentication-concept.adoc[Web authentication using OIDC authorization code flow]
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-authentication-mechanisms-concept.adoc:157]: I'm not sure about `Web authentication using OIDC authorization code flow`, OIDC is what we use in the docs as an acronym, but the `Using OpenID Connect to Protect Web Applications` is precisely what the role of this code flow mechanism is. `Web authentication` is not really a correct term when it applies to Quarkus. Typing it in Google leads to WebAuthn docs among other things.

So please use the original `Using OpenID Connect to Protect Web Applications` or if you'd like, may be something like `OpenId Connect authorization code flow to protect web applications` ?

File: docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc [hunk 9]
```diff
@@ -0,0 +1,264 @@
+[id="security-oidc-protect-web-application-tutorial"]
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc:1]: Please rename it as `security-oidc-web-authentication-tutorial`

File: docs/src/main/asciidoc/security-oidc-web-authentication-concept.adoc.adoc [hunk 10]
```diff
@@ -0,0 +1,1239 @@
+////
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-oidc-web-authentication-concept.adoc.adoc:1]: double `.adoc` is here

File: docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc [hunk 11]
```diff
@@ -7,1458 +7,3 @@ https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
 include::_attributes.adoc[]
 :categories: security
 :summary: This guide demonstrates how to use the OpenID Connect Extension to protect your web application based on the Authorization Code Flow using Quarkus.
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc:9]: I'm not sure I understand what is the role of this document now, it is empty. I thought it could make sense to actually retain the introduction with the diagram here and then link to 1) tutorial 2) reference material

File: docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc [hunk 12]
```diff
@@ -0,0 +1,264 @@
...
+
+After clicking the `Login` button, you are redirected back to the application.
+
+For more information about writing the integration tests that depend on `Dev Services for Keycloak`, see the <<integration-testing-keycloak-devservices, Dev Services for Keycloak>> section.
+
+== Summary
+
+Congratulations!
+You have learned how to set up and use the OIDC authorization code flow mechanism to protect and test application HTTP endpoints.
+After you have completed this tutorial, explore some of the other security mechanisms in Quarkus.
```
Comment (sberyozkin) [inline docs/src/main/asciidoc/security-oidc-protect-web-application-tutorial.adoc:249]: I think the line 249 is not necessary, and the link to the bearer token concept should be in References, as well as the security concept one. 
As far as the next steps in scope of this tutorial is concerned, it should be the web-app reference (concept) and OIDC multi-tenency

File: docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc (hunk 11 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-openid-connect-web-authentication.adoc:9]: OK, I see it has been removed

File: docs/src/main/asciidoc/security-oidc-web-authentication-concept.adoc.adoc (hunk 10 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-web-authentication-concept.adoc.adoc:1]: I'm not sure now about this restructuring. So the remove oidc web-authentication doc had this structure:
- Introduction with the nice diagram you prepared
- Tutorial/Quickstart
- Reference

Now we have 
- Tutorial in a separate doc
- Introduction with the diagram combined with the all the reference material

I'm not sure it makes it better for the users.
I'd like to propose 2 options:
1. Move the Intro with the diagram to the top of the tutorial doc - so that by the time they go to the tutorial they have a high level picture in mind
2. Restore the original doc, keep the intro there and link to the tutorial and concept

Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-web-authentication-concept.adoc.adoc:1]: Note security-authentication-mechanisms link to this doc but neither of these docs refer to the tutorial. IMHO finding the tutorial should be super easy. I propose to have both of these docs refer to it. This doc should offer the tutorial link at the very top and in the references IMHO

**Comment by sberyozkin** [comment]
Good evening @sheilamjones :-), thanks for this PR, a lot of work has been done here which is very appreciated.
IMHO we need to optimize a few things here though, lets discuss it in the next day or so.
Cheers

**Comment by sberyozkin** [comment]
By the way, Sheila, your idea to have `authorization code` reflected somehow in the doc names is nice, I thought about it this morning, the current base name `security-openid-connect-web-authentication-*.adoc` does not reflect it, and I guess `web-authentication` part in this name can cause some confusion. But FYI, `security-openid-connect.adoc` does not reflect it either it is about a bearer authentication.
Lets agree on the restructuring aspects first and then if you don't mind, perhaps we should indeed follow our lead and rename these new docs, perhaps `security-openid-connect-web-authentication-*.adoc` -> `security-openid-connect-web-authentication-*.adoc` -> `security-openid-connect-authorization-code-*.adoc` or something like that

**Comment by sberyozkin** [comment]
Made a bad typo, it was your lead, I typed `our` :-), not sure how I managed to type it but fixed now :-)

File: docs/src/main/asciidoc/security-oidc-web-authentication-concept.adoc.adoc (hunk 10 above)
Comment (sberyozkin) [reply docs/src/main/asciidoc/security-oidc-web-authentication-concept.adoc.adoc:1]: Perhaps it is fine to keep the diagramm in the concept but the users coming to the tutorial should be asked to check it and the whole intro first, some brief intro at the top of the tutorial might be worth having as well

**Comment by sberyozkin** [comment]
@sheilamjones Excellent work, thanks, we might tweak a few more things going forward, but it does look good now, cheers

**Review by sberyozkin — APPROVED** [review]
